state.go

```diff
@@ -422,17 +452,26 @@ type AuthorizedState struct {
 	BaseState
 }
 
-func (a *AuthorizedState) Handle(c Controller) (State, bool) {
-	// TODO HasUpgrade should return update information
-	has, err := c.HasUpgrade()
+func (a *AuthorizedState) Handle(ctx *StateContext, c Controller) (State, bool) {
+	sd, err := LoadStateData(ctx.store)
 	if err != nil {
```

Comment:
Couple of things here. First I think it might have sense to consider update as failed if the state can not be retrieved. 
As for the logic:
```
lastKnownState, imageID := LoadStateData()
if err != ErrNotExist
  if HasUpgrade()
    // probably something bad happened during update
    send broken update to server // might be tricky as we don't have update id
  else
    log error
    start from scratch?
// we have installed update correctly (can we enter this state if not?)
swith lastKnownState 
  case MenderStateReboot
    if HasUpgrade()
      if curentInamgeID == imageID
        update complete && report success
      else
        send broken update to server with imageID
    else
      last state was reboot but no update; is this valid?

  case MenderStateAuthorized
    do nothing; we simply rebooted

  default
    // update was in progress, but something went wrong
    report broken update to server with imageID
    start from scratch

RemoveStateData()
```